File: examples/runtime/openai_chat.py
#!/usr/bin/env python3
"""Chat with a locally served JANG/JANGTQ bundle.

Start the server first:
    JANG_MODEL=/path/to/model ./examples/runtime/serve_vmlx.sh
"""

from __future__ import annotations

import argparse
import os
import sys

from openai import OpenAI


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--base-url", default=os.getenv("OPENAI_BASE_URL", "http://127.0.0.1:8000/v1"))
    parser.add_argument("--api-key", default=os.getenv("OPENAI_API_KEY", "not-needed"))
    parser.add_argument("--model", default=os.getenv("JANG_MODEL_ALIAS", "default"))
    parser.add_argument("--prompt", default="Explain JANG in one paragraph.")
    parser.add_argument("--system", default="You are a precise technical assistant.")
    parser.add_argument("--max-tokens", type=int, default=256)
    parser.add_argument("--temperature", type=float, default=0.0)
    parser.add_argument("--stream", action="store_true")
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    client = OpenAI(base_url=args.base_url, api_key=args.api_key)
    messages = [
        {"role": "system", "content": args.system},
        {"role": "user", "content": args.prompt},
    ]

    if args.stream:
        stream = client.chat.completions.create(
            model=args.model,
            messages=messages,
            max_tokens=args.max_tokens,
            temperature=args.temperature,
            stream=True,
        )
        for chunk in stream:
            delta = chunk.choices[0].delta
            reasoning = getattr(delta, "reasoning", None) or getattr(delta, "reasoning_content", None)
            if reasoning:
                print(reasoning, end="", file=sys.stderr, flush=True)
            if delta.content:
                print(delta.content, end="", flush=True)
        print()
        return 0

    response = client.chat.completions.create(
        model=args.model,
        messages=messages,
        max_tokens=args.max_tokens,
        temperature=args.temperature,
    )
    msg = response.choices[0].message
    reasoning = getattr(msg, "reasoning", None) or getattr(msg, "reasoning_content", None)
    if reasoning:
        print("Reasoning:")
        print(reasoning)
        print()
    print(msg.content or "")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: examples/runtime/swift/OpenAIChat.swift
import Foundation

struct ChatRequest: Encodable {
    let model: String
    let messages: [Message]
    let max_tokens: Int
    let temperature: Double
}

struct Message: Codable {
    let role: String
    let content: String
}

struct ChatResponse: Decodable {
    struct Choice: Decodable {
        let message: Message
    }

    let choices: [Choice]
}

func value(after flag: String, in args: [String], default fallback: String) -> String {
    guard let index = args.firstIndex(of: flag), index + 1 < args.count else {
        return fallback
    }
    return args[index + 1]
}

let args = CommandLine.arguments
let baseURL = value(after: "--base-url", in: args, default: "http://127.0.0.1:8000/v1")
let apiKey = value(after: "--api-key", in: args, default: "not-needed")
let model = value(after: "--model", in: args, default: "default")
let prompt = value(after: "--prompt", in: args, default: "Explain JANG in one paragraph.")

let requestBody = ChatRequest(
    model: model,
    messages: [
        Message(role: "system", content: "You are a precise technical assistant."),
        Message(role: "user", content: prompt),
    ],
    max_tokens: 256,
    temperature: 0.0
)

let endpoint = URL(string: baseURL.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/chat/completions")!
var request = URLRequest(url: endpoint)
request.httpMethod = "POST"
request.setValue("application/json", forHTTPHeaderField: "Content-Type")
request.setValue("Bearer \(apiKey)", forHTTPHeaderField: "Authorization")
request.httpBody = try JSONEncoder().encode(requestBody)

let semaphore = DispatchSemaphore(value: 0)
let task = URLSession.shared.dataTask(with: request) { data, response, error in
    defer { semaphore.signal() }

    if let error {
        fputs("request failed: \(error)\n", stderr)
        return
    }

    guard let http = response as? HTTPURLResponse else {
        fputs("missing HTTP response\n", stderr)
        return
    }

    guard (200..<300).contains(http.statusCode), let data else {
        let body = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""
        fputs("HTTP \(http.statusCode): \(body)\n", stderr)
        return
    }

    do {
        let decoded = try JSONDecoder().decode(ChatResponse.self, from: data)
        print(decoded.choices.first?.message.content ?? "")
    } catch {
        fputs("decode failed: \(error)\n", stderr)
    }
}

task.resume()
semaphore.wait()
